CORE-2928: Comparable bloom filters without template copy

mauguignard opened the pull request:
As reported by @gonzalezzfelipe, the idea of this ticket is to support `union` and `intersection` between two different BloomFilters created _equal_. Since the `hash_seeds` are created randomly when initializing the structure, I added an optional `seed` argument to configure the random number generator of the `random` module.

However, since we copy the whole BloomFilter structure to the MBArray header, the `_assert_comparable` silently failed and prevented us from using the `union` and `intersection` methods even on two BloomFilters created with the same parameters.

For more info about this ticket please checkout [JIRA](https://jampphq.atlassian.net/browse/CORE-2928).

Commit:
CORE-2928: Allow to create comparable bloom filters without copying a template

llogiudice commented on src/bloomfilter.c:
```diff
@@ -18,22 +18,20 @@ BloomFilter *bloomfilter_Create_Malloc(size_t max_num_elem, double error_rate,
+    bf->header.error_rate = error_rate;
+    bf->header.num_hashes = num_hashes;
+    bf->header.bf_version = BF_CURRENT_VERSION;
+    memset(bf->header.hash_seeds, 0, sizeof(uint32_t) * 256);
```
Don't use hardcoded constants. Use `sizeof (bf->header.hash_seeds)` if you're clearing a whole member.

Better yet: Don't zero `header.hash_seeds` if you're going to (partially) overwrite it in the next line.

mauguignard replied:
I limited myself to rename `bf->hash_seeds` to `bf->header.hash_seeds`, the hardcoded constant was already there. However, I agree, so I have fixed it.

On the other hand, the `memset` *is* necessary, otherwise, we would leave garbage in the `hash_seeds` array that would break the `_assert_comparable` results. And fix that is precisely the objective of this PR.

llogiudice replied:
That's not what I meant. Instead of clearing a full array and then copying over it, you can do just what's needed:

```
memset (mempcpy (bf->header.hash_seeds, hash_seeds, sizeof (uint32_t) * num_hashes),
                 0, sizeof (bf->header.hash_seeds) - sizeof (uint32_t) * num_hashes);
```

But it's a nitpick. OTOH, since many of the field members are zero-initialized, you can get away with using `calloc` to allocate storage for the BF. But again, it's just a nitpick. I pointed out the usage of a hardcoded constant because it's really bad practice in C (this whole library is a mess, anyway).

mauguignard replied:
Done.

llogiudice commented on src/bloomfilter.c:
```diff
@@ -72,14 +66,18 @@ BloomFilter *bloomfilter_Create_Mmap(size_t max_num_elem, double error_rate,
        back into this BloomFilter object.
     */
-    if (mbarray_Header((char *)bf, array, sizeof(BloomFilter)) == NULL) {
+    if (mbarray_Header((char *)&bf->header, array, sizeof(Header)) == NULL) {
```
Have you checked that structs `Header` and `Bloomfilter` are binary compatible ? If they aren't - Doesn't this mean that bloomfilters compiled with 0.3.16j aren't usable in 0.3.17j ?

mauguignard replied:
They are now compatible.

klaussfreire commented on src/bloomfilter.h:
```diff
@@ -6,26 +6,29 @@
-    MBArray * array;
     unsigned char bf_version;
+
+    uint32_t hash_seeds[256];
```
You're moving stuff around. You can't move stuff around without breaking binary compatibility with existing BFs.

While it may make the act of copying fields over a tad more involved, I think you should do this without changing the structure at all.

mauguignard replied:
I needed to move some fields around to remove the problematic fields from the header. Adding an `unused` field and modifying the `reserved` array size to compensate for the removed fields I was able to restore the binary compatibility.

klaussfreire replied:
No, you have to leave the structure alone to be able to restore binary compatibility.

It's not just the size of the struct, it's the location of fields in the struct. These files are memory-mapped and read from file, if you move things around, the earlier BFs are unreadable.

klaussfreire replied:
Sorry... I was confused. The struct that's stored in file is another struct (MmapBitArray), so this one seems internal and you can play with it however you like.

So... go ahead and destroy binary compatibility if you want ;)

klaussfreire replied:
Disregard the deleted message. This module is really ugly and unclear about what it does. It seems it steps onto the first few elements of the BloomFilter struct, so you must keep it compatible. But only the first few.

mauguignard replied:
Hahaha, either way I modified the `Header` struct to be the same as `_BloomFilter` was (I just renamed the unused fields to make clear that they are now deprecated to avoid potential bugs). I have checked and I can open bloomfilters created with the previous version just fine and the values are as expected. Could you please review it again to make sure that I'm not missing something? Thanks!

klaussfreire replied:
Ok, yes. There's a lot of black magic involved. I would leave the data structures alone, odds are you'll break something if you don't. This is not clear code.

klaussfreire replied:
> Hahaha, either way I modified the `Header` struct to be the same as `_BloomFilter` was (I just renamed the unused fields to make clear that they are now deprecated to avoid potential bugs). I have checked and I can open bloomfilters created with the previous version just fine and the values are as expected. Could you please review it again to make sure that I'm not missing something? Thanks!

Yeahno. I'm absolutely sure you don't need to change anything in the data structures to implement this. So if you change something, it's wrong.

You just have to have an option to create 2 independent BFs with the same seeds. That's all.

klaussfreire commented on src/bloomfilter.h:
```diff
@@ -6,26 +6,29 @@
+
     unsigned char count_correct;
     uint64_t elem_count;
-    uint32_t reserved[32];
```
That includes the reserved space.

mauguignard replied:
Restored!

Commit:
CORE-2928: Fix binary compatibility

Commit:
CORE-2928: Fix missing NULL set

llogiudice commented:
The way you "restored" binary compatibility seems a bit iffy to me. It appears that you just checked that the sizes of the old and new structure are equal and left it at that. But reordering struct members can have an impact on alignment, or member offset assignment. In order to avoid subtle bugs (remember that the bf header is being `pread` from a file), I recommend keeping the structure as it was, and then replicating the additional members `count_correct`, `elem_count`, and `array` outside of the header.

Commit:
CORE-2928: Restore structure to assure binary compatibility

Commit:
CORE-2928: Use calloc instead of malloc

mauguignard commented:
> The way you "restored" binary compatibility seems a bit iffy to me. It appears that you just checked that the sizes of the old and new structure are equal and left it at that. But reordering struct members can have an impact on alignment, or member offset assignment. In order to avoid subtle bugs (remember that the bf header is being `pread` from a file), I recommend keeping the structure as it was, and then replicating the additional members `count_correct`, `elem_count`, and `array` outside of the header.

Done.

Commit:
CORE-2928: Use better naming to avoid potential bugs on deprecated fields

klaussfreire commented:
> As reported by @gonzalezzfelipe, the idea of this ticket is to support `union` and `intersection` between two different BloomFilters created _equal_. Since the `hash_seeds` are created randomly when initializing the structure, I added an optional `seed` argument to configure the random number generator of the `random` module.
> 
> However, since we copy the whole BloomFilter structure to the MBArray header, the `_assert_comparable` silently failed and prevented us from using the `union` and `intersection` methods even on two BloomFilters created with the same parameters.
> 
> The `count_correct`, `elem_count` and `array` fields could (and in fact, we should expect them to) be different between two BloomFilters and they should still be comparable.

Not sure what you're talking about.

All `_assert_comparable` checks is bits and hash_seeds. If you generate the seeds the same, and bits ends up being the same, the BFs are comparable.

You absolutely don't need to change anything else. You did something else wrong if you're forced to change the data structures.

mauguignard commented:
> > As reported by @gonzalezzfelipe, the idea of this ticket is to support `union` and `intersection` between two different BloomFilters created _equal_. Since the `hash_seeds` are created randomly when initializing the structure, I added an optional `seed` argument to configure the random number generator of the `random` module.
> > However, since we copy the whole BloomFilter structure to the MBArray header, the `_assert_comparable` silently failed and prevented us from using the `union` and `intersection` methods even on two BloomFilters created with the same parameters.
> > The `count_correct`, `elem_count` and `array` fields could (and in fact, we should expect them to) be different between two BloomFilters and they should still be comparable.
> 
> Not sure what you're talking about.
> 
> All `_assert_comparable` checks is bits and hash_seeds. If you generate the seeds the same, and bits ends up being the same, the BFs are comparable.
> 
> You absolutely don't need to change anything else. You did something else wrong if you're forced to change the data structures.

The problem is that `_assert_comparable` checks more than that (that's why it silently fails), it checks the whole preamble (basically, what `_initialize_file` writes). Since the preamble includes the _header_, and that was the previous `BloomFilter` struct dump, it included all the fields in the struct, not just `hash_seeds`. So `_assert_comparable` was more demanding than it should since it checked that `count_correct`, `elem_count` and even the memory address pointed by `array` were exactly the same in order to be _comparables_. So basically, `_assert_comparable` checked if both bloom filters were equal, that's the reason why I removed those fields from the header.

klaussfreire commented:
> > > As reported by @gonzalezzfelipe, the idea of this ticket is to support `union` and `intersection` between two different BloomFilters created _equal_. Since the `hash_seeds` are created randomly when initializing the structure, I added an optional `seed` argument to configure the random number generator of the `random` module.
> > > However, since we copy the whole BloomFilter structure to the MBArray header, the `_assert_comparable` silently failed and prevented us from using the `union` and `intersection` methods even on two BloomFilters created with the same parameters.
> > > The `count_correct`, `elem_count` and `array` fields could (and in fact, we should expect them to) be different between two BloomFilters and they should still be comparable.
> > 
> > 
> > Not sure what you're talking about.
> > All `_assert_comparable` checks is bits and hash_seeds. If you generate the seeds the same, and bits ends up being the same, the BFs are comparable.
> > You absolutely don't need to change anything else. You did something else wrong if you're forced to change the data structures.
> 
> The problem is that `_assert_comparable` checks more than that (that's why it silently fails), it checks the whole preamble (basically, what `_initialize_file` writes). Since the preamble includes the _header_, and that was the previous `BloomFilter` struct dump, it included all the fields in the struct, not just `hash_seeds`.

Um... no it doesn't:

``` python
    def _assert_comparable(self, BloomFilter other):
        error = ValueError("The two %s objects are not the same type (hint, "
                           "use copy_template)" % self.__class__.__name__)
        if self._bf.array.bits != other._bf.array.bits:
            raise error
        if self.hash_seeds != other.hash_seeds:
            raise error
        return
```

If `hash_seeds` is initialized correctly that is. Maybe there's a bug there.

mauguignard commented:
> Um... no it doesn't:
> 
> ```python
>     def _assert_comparable(self, BloomFilter other):
>         error = ValueError("The two %s objects are not the same type (hint, "
>                            "use copy_template)" % self.__class__.__name__)
>         if self._bf.array.bits != other._bf.array.bits:
>             raise error
>         if self.hash_seeds != other.hash_seeds:
>             raise error
>         return
> ```
> 
> If `hash_seeds` is initialized correctly that is. Maybe there's a bug there.

Oh! I see were all the confusion is coming from. I was refering to this `_assert_comparable`: https://github.com/jampp/pybloomfiltermmap/blob/master/src/mmapbitarray.c#L386

(Maybe having two different functions with the same name was not the library's author's best idea :sweat_smile: ).

klaussfreire commented:
> Oh! I see were all the confusion is coming from. I was refering to this `_assert_comparable`: https://github.com/jampp/pybloomfiltermmap/blob/master/src/mmapbitarray.c#L386
> 
> (Maybe having two different functions with the same name was not the library's author's best idea sweat_smile ).

I see. Why  not fix that one?

mauguignard commented:
> I see. Why not fix that one?

The problem is that the MBArray is completely ignorant about what the contents of the `header` are. It compares the preamble as a whole and we don't have a way of telling it which bits it should consider and which it should not. On the other hand, we could remove that check, but I don't think that's a proper solution since we don't want to allow the union/intersection between two different bloom filters with a different `max_num_elem` or `error_rate` (as you said, we currently check for the number of `bits` and the `hash_seeds` in the Cython part).

On the other hand, the earlier discussion got me thinking and now I realize that the bug is not actually in the `count_correct`, `elem_count` nor `array` fields, since we always used:

```c
    bf->count_correct = 1;
    bf->elem_count = 0;
    bf->array = NULL;
```

when initializing the structure before creating the preamble. And as reported in [CORE-2946](https://jampphq.atlassian.net/browse/CORE-2946) we never update it afterward, so those values are not the problem. Give me a second, I think I found a simple solution that doesn't require to modify the structure and will leave everyone happy (until we try to fix CORE-2946 at least).

klaussfreire commented:
> > I see. Why not fix that one?
> 
> The problem is that the MBArray is completely ignorant about what the contents of the `header` are. It compares the preamble as a whole and we don't have a way of telling it which bits it should consider and which it should not.

It should consider none. The higher-level structure is the one responsible for checking the preamble.

The MBArray should only check that the bitmap size and offset is comparable between both (preamble bytes, array bits, type, etc)

> On the other hand, we could remove that check, but I don't think that's a proper solution since we don't want to allow the union/intersection between two different bloom filters with a different `max_num_elem` or `error_rate` (as you said, we currently check for the number of `bits` and the `hash_seeds` in the Cython part).

That won't happen. The higher-level `assert_comparable` structure would simply check that.

Commit:
CORE-2928: Rollback all changes

Commit:
CORE-2928: Re-add seed parameter

gonzalezzfelipe commented:
> > > I see. Why not fix that one?
> > 
> > 
> > The problem is that the MBArray is completely ignorant about what the contents of the `header` are. It compares the preamble as a whole and we don't have a way of telling it which bits it should consider and which it should not.
> 
> It should consider none. The higher-level structure is the one responsible for checking the preamble.
> 
> The MBArray should only check that the bitmap size and offset is comparable between both (preamble bytes, array bits, type, etc)
> 
> > On the other hand, we could remove that check, but I don't think that's a proper solution since we don't want to allow the union/intersection between two different bloom filters with a different `max_num_elem` or `error_rate` (as you said, we currently check for the number of `bits` and the `hash_seeds` in the Cython part).
> 
> That won't happen. The higher-level `assert_comparable` structure would simply check that.

Just here to put my two cents on this, when I made this ticket I simply removed the header check to see if that was the problem, and everything worked just fine. I agree with @klaussfreire in that the compatibility check is already being done in the higher-level `assert_comparable`, and if it were up to my I would simply remove that check.

PD: By _that check_ I mean changing the `_assert_comparable` function on `mmapbitarray.c` from:
```c
static inline int _assert_comparable(MBArray * array1, MBArray * array2)
{
    errno = EINVAL;
    if (array1->preamblebytes != array2->preamblebytes) {
        return 1;
    }

    if (memcmp((char *)array1->vector, (char *)array2->vector, array1->preamblebytes)) {
        return 1;
    }

    return 0;
}
```
to
```c
static inline int _assert_comparable(MBArray * array1, MBArray * array2)
{
    errno = EINVAL;
    if (array1->preamblebytes != array2->preamblebytes) {
        return 1;
    }
    return 0;
}
```
This wont check that the seeds are the same but at least it checks for the amount there is 🤷‍♂

Commit:
CORE-2928: Add some tests

Commit:
CORE-2928: Fix bug: The structure should use calloc for initialization

Commit:
CORE-2928: Bump library version

Commit:
CORE-2928: Improve Cython _assert_comparable

Commit:
CORE-2928: Remove header from MBArray _assert_comparable

mauguignard commented:
As discussed in the SE Weekly the bug was not in the `_assert_comparable` per se. The bug was in how the struct was declared and how it was initialized.
As previously discussed, struct members' memory addresses are affected by the padding added by the compiler to comply with the alignment expected by the platform.

Since the struct was initialized used `malloc` it left garbage in the 2 unreachable bytes between `count_correct` and `elem_count`. That would normally not be a problem except for the fact that we are dumping (and afterward comparing) the struct as a whole, including those 2 bytes that were never cleared.

Using `calloc` instead of `malloc` should solve the issue as you can verify for yourselves in [85e3e6cb]. It should also be faster since we can now save calling `memset` to clear the `hash_seeds` and `reserved` arrays.

That solution, however, doesn't take into account the BloomFilters created on previous versions of this library. To solve that issue, as proposed by @klaussfreire and @gonzalezzfelipe I reinforced the high-level `_assert_comparable` to check not only for the number of `bits` in the array and the `hash_seeds`, but for all the properties of the structure.

After that, I removed the check for the header contents in the low-level `_assert_comparable` (It now only checks that the file has the expected _magic_, is of the expected size and the `header_len`s are equal).

klaussfreire commented on src/pybloomfilter.pyx:
```diff
@@ -143,6 +143,8 @@ cdef class BloomFilter:
             #    ** num_hashes)
 
+            if seed is not None:
+                random.seed(seed)
```
Use a local random state instance, not the global random state.

mauguignard replied:
Done!

klaussfreire commented on src/pybloomfilter.pyx:
```diff
@@ -351,8 +353,9 @@ cdef class BloomFilter:
             raise error
-        if self.hash_seeds != other.hash_seeds:
-            raise error
+        for prop in ('capacity', 'error_rate', 'num_hashes', 'num_bits', 'hash_seeds'):
```
You already checked num_bits above

mauguignard replied:
Ups! Fixed!

Commit:
CORE-2928: Fix duplicated check for num_bits

Commit:
CORE-2928: Use local random instance on __cinit__

llogiudice commented on src/mmapbitarray.c:
```diff
@@ -390,7 +390,7 @@ static inline int _assert_comparable(MBArray * array1, MBArray * array2)
     }
 
-    if (memcmp((char *)array1->vector, (char *)array2->vector, array1->preamblebytes)) {
+    if (memcmp((char *)array1->vector, (char *)array2->vector, MBAMAGICSIZE + sizeof(BTYPE) + sizeof(int32_t))) {
```
This doesn't make much sense. The amount of bytes you're comparing is _19_ (9 + 8 + 4), which isn't a multiple of the element size.

What exactly are you trying to achieve here ?

mauguignard replied:
It checks that the first 9 bytes are equal (they both should be "MBITARRAY"). The next 8 bytes contains a `uint64_t` that indicates the number of bits of the bit array (`num_bits` in the Cython part). And the following 4 bytes contains a `int32_t` that indicates the number of bytes of the header.

I'm basically comparing what `_initialize_file` writes, skipping the header itself and the ending null character that I just don't care:

```c
int _initialize_file(int fd, size_t end, BTYPE num_bits, const char * header, int32_t header_len)
{
    unsigned char zero = 0;
    errno = 0;
    lseek(fd, 0, SEEK_SET);
    if (write(fd, MBAMAGIC, MBAMAGICSIZE) != MBAMAGICSIZE) {
        return 1;
    }
    if (write(fd, &num_bits, sizeof(BTYPE)) != sizeof(BTYPE)) {
        return 1;
    }
    if (write(fd, &header_len, sizeof(header_len)) != sizeof(header_len)) {
        return 1;
    }
    if (header_len) {
        if (write(fd, header, header_len) != header_len) {
            return 1;
        }
    }

    lseek(fd, end, SEEK_SET);
    if (write(fd, &zero, 1) != 1) {
        return 1;
    }

    if (errno) {
        return 1;
    }
    return 0;
}
```

llogiudice replied:
Ugh. This is absolutely sloppy. The header should be in a byte buffer instead of embedded directly in the vector. But whatever, this is not the PR for that.
